### PR TITLE
[file system][ios] Add multipart upload

### LINKED ISSUE #5841: [iOS] Look into implementing Background Upload
Using [NSURLSessionUploadTask](https://developer.apple.com/documentation/foundation/nsurlsessionuploadtask?language=objc)

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Part of https://github.com/expo/expo/issues/5841.

# How

- add multipart upload on iOS

# Test Plan

- https://snack.expo.io/@lukaszkosmaty/expo-file-system---add-multipart-upload ✅

### CODE REVIEW

File: packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m [hunk 1]
```diff
@@ -651,6 +668,51 @@ - (NSDictionary *)encodingMap
 
 #pragma mark - Internal methods
 
+// Borrowed from http://stackoverflow.com/questions/2439020/wheres-the-iphone-mime-type-database
+- (NSString *)_guessMIMETypeFromPath:(NSString *)path {
```
Comment (tsapeta) [inline packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m]:
```suggestion
- (NSString *)_guessMIMETypeFromPath:(NSString *)path
{
```

File: packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m [hunk 2]
```diff
@@ -715,6 +777,15 @@ - (NSString *)_importHttpMethod:(NSNumber *)httpMethod
   return @"POST";
 }
 
+- (BOOL)_isRawUploadType:(NSNumber * _Nullable)uploadType
+{
+  if ([uploadType isEqual:@1]) {
+    return false;
+  }
+  
+  return true;
```
Comment (tsapeta) [inline packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m]:
```suggestion
  return ![uploadType isEqual:@1])
```
? 😄

**Review by tsapeta — CHANGES_REQUESTED** [review]
Missing docs and changelog entry 😉

File: packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m [hunk 3]
```diff
@@ -651,6 +668,52 @@ - (NSDictionary *)encodingMap
...
+                         parameters:(NSDictionary * _Nullable)parameters
+                          fieldName:(NSString * _Nullable)fieldName
+                           mimeType:(NSString * _Nullable)mimetype
+{
+
+  NSMutableData *body = [NSMutableData data];
+
+  NSData *data = [NSData dataWithContentsOfURL:fileUri];
+  
+  NSString *filename  = [[fileUri path] lastPathComponent];
```
Comment (bbarthec) [inline packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m]:
```suggestion
  NSMutableData *body = [NSMutableData data];
  NSData *data = [NSData dataWithContentsOfURL:fileUri];
  NSString *filename  = [[fileUri path] lastPathComponent];
```
Why so many empty lines 🤔 ?

File: packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m [hunk 4]
```diff
@@ -555,12 +556,28 @@ - (NSDictionary *)encodingMap
...
+  } else {
+    NSString *boundaryString = [[NSUUID UUID] UUIDString];
+    [urlRequest setValue:[NSString stringWithFormat:@"multipart/form-data; boundary=%@", boundaryString] forHTTPHeaderField:@"Content-Type"];
+
+    NSData *httpBody = [self _createBodyWithBoundary:boundaryString
+                                             fileUri:fileUri
+                                          parameters:options[@"parameters"]
+                                           fieldName:options[@"fieldName"]
+                                            mimeType:options[@"mimeType"]];
+    [urlRequest setHTTPBody: httpBody];
```
Comment (bbarthec) [inline packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m]:
```suggestion
    [urlRequest setHTTPBody:httpBody];
```
💅

File: packages/expo-file-system/src/FileSystem.types.ts [hunk 5]
```diff
@@ -3,6 +3,11 @@ export enum FileSystemSessionType {
   FOREGROUND = 1,
 }
 
+export enum FileSystemUploadType {
+  RAW = 0,
```
Comment (bbarthec) [inline packages/expo-file-system/src/FileSystem.types.ts]: This name is borrowed from some web standard? 🤔

File: packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m [hunk 6]
```diff
@@ -702,16 +761,30 @@ - (NSMutableDictionary *)_parseServerResponse:(NSURLResponse *)response
...
-  }
-  
-  return @"POST";
+
+  return [httpMethod isEqual:@(EXFileSystemPatchMethod)] ? @"PATCH" : @"POST";
+}
+
+- (BOOL)_isRawUploadType:(NSNumber * _Nullable)uploadType
+{
+  return ![uploadType isEqual:@1];
```
Comment (bbarthec) [inline packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m]: I dislike this one .. 😞 
`1)` I saw that you decided on naming it differently than `RAW` (`BINARY_CONTENT` 🤔)
`2)` This magic number `1` is like an eyesore 😭

**Review by bbarthec — CHANGES_REQUESTED** [review]
LGTM 💯 But one little thing 😉

File: packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m [hunk 7]
```diff
@@ -703,6 +730,61 @@ - (NSDictionary *)encodingMap
 
 #pragma mark - Internal methods
 
+- (EXFileSystemUploadType)_importUploadType:(NSNumber * _Nullable)type
```
Comment (bbarthec) [inline packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m:733]:
```suggestion
- (EXFileSystemUploadType)getUploadTypeFrom:(NSNumber * _Nullable)type
```
Semantically what you're doing here is just getting import type from some integer value with fallback when value is not in available ENUM values.

File: packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m [hunk 8]
```diff
@@ -703,6 +730,61 @@ - (NSDictionary *)encodingMap
...
+{
+  switch ([type intValue]) {
+    case EXFileSystemBinaryContent:
+      return EXFileSystemBinaryContent;
+    case EXFileSystemMultipart:
+      return EXFileSystemMultipart;
+  }
+  
+  return EXFileSystemInvalidType;
```
Comment (bbarthec) [inline packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m:742]: That looks terrible TBH ... 😉 
I've googled a bit and found out some other tricky and not-so-pretty way:
```suggestion
  switch ([type intValue]) {
    case EXFileSystemBinaryContent:
    case EXFileSystemMultipart:
      return [type intValue];
  }
  
  return EXFileSystemInvalidType;
```

**Review by bbarthec — APPROVED** [review]
In general it LGTM 😉 💯 Good job. Left some minor comments.

**Pull request merged**